### ISSUE TITLE
finitevolx (optional/low-priority): Full Multigrid (FMG) cycle + differentiable min helpers

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#229]
## TL;DR

An audit against the MASSH SSH-mapping codebase ([`leguillf/MASSH@VarDyn`](https://github.com/leguillf/MASSH/tree/VarDyn)) found that **finitevolx is already a superset of MASSH's numerical primitives** — finite differences, flux/WENO reconstruction, masks, and notably the **variable-coefficient Helmholtz multigrid** (`finitevolx/_src/solvers/multigrid.py:1` — *"Geometric multigrid solver for variable-coefficient Helmholtz equations"*, with `_interpolate_coeff_to_faces`, mask hierarchies, and `v_cycle`).

So there is **nothing required** here. This issue tracks two **optional, low-priority** niceties that MASSH has and finitevolx does not:

1. **Full Multigrid (FMG)** cycle on top of the existing V-cycle.
2. **Differentiable min/abs helpers** (`smooth_abs`, `smooth_clamp`) for adjoint-stable WENO / positive-thickness clamping.

---

## Background & motivation

MASSH's `model_qgsw/helmholtz_multigrid.py` provides `class MG_Helmholtz` with `V_cycle` **and** `FMG_Helmholtz` (Full Multigrid: build the solution from the coarsest level up, nesting V-cycles). FMG gives a better initial guess and can reach discretization-error accuracy in ~one nested sweep — useful when the elliptic solve dominates cost (PV inversion every RHS evaluation). finitevolx's `MultigridSolver` exposes `v_cycle` / `solve_onestep` / `solve_unrolled` but no FMG wrapper.

Separately, MASSH `model_qgsw/sw.py` and `reconstruction.py` use differentiable surrogates to keep the adjoint well-behaved (verbatim):

- `def smooth_abs(x):` — `√(x² + ε²)`, a differentiable `|x|` (used in WENO smoothness indicators).
- `def smooth_clamp(x, x_min, sharpness=10.):` — softplus-based differentiable `max(x, x_min)` (keeps layer thickness positive without a hard, non-differentiable clamp).

finitevolx targets differentiable ocean models (JAX), so these are natural small additions; today a user hand-rolls them.

---

## User stories

- **As someone solving the same elliptic problem every time step**, I want an FMG entry point so PV inversion converges to discretization accuracy in roughly one nested sweep.
- **As a builder of adjoint-stable WENO / positive-definite-thickness models**, I want `smooth_abs` / `smooth_clamp` as documented, tested helpers rather than re-deriving the `ε`/sharpness each time.

---

### 1. FMG (`finitevolx/_src/solvers/multigrid.py`, extending `MultigridSolver`)

```python
class MultigridSolver(eqx.Module):
    ...
    def fmg(self, rhs: Float[Array, "Ny Nx"], *, n_vcycles: int = 1) -> Float[Array, "Ny Nx"]:
        """Full Multigrid: restrict rhs to coarsest level, solve, then prolong
        and refine with `n_vcycles` V-cycles per level up to the finest grid.
        Mirrors MASSH MG_Helmholtz.FMG_Helmholtz."""
```

### 2. Differentiable helpers (`finitevolx/_src/utils/` or alongside reconstruction)

```python
def smooth_abs(x: Array, eps: float = 1e-12) -> Array:
    """Differentiable |x| ~ sqrt(x**2 + eps**2). Mirrors MASSH smooth_abs."""

def smooth_clamp(x: Array, x_min: float, sharpness: float = 10.0) -> Array:
    """Differentiable max(x, x_min) via softplus. Mirrors MASSH smooth_clamp."""
```

---

## API examples

```python
from finitevolx import build_multigrid_solver

solver = build_multigrid_solver(grid=grid, mask=mask, coeff=rossby_radius_sq, lam=0.0)
psi = solver.fmg(rhs, n_vcycles=1)        # discretization-accuracy in one nested sweep
```

```python
from finitevolx import smooth_clamp
h_pos = smooth_clamp(h_total, x_min=h_floor, sharpness=10.0)   # keep thickness > 0, differentiably
```

---

## TODO

- [ ] `MultigridSolver.fmg(rhs, n_vcycles=1)` reusing the existing restriction/prolongation/level hierarchy.
- [ ] `smooth_abs`, `smooth_clamp` helpers + public export.
- [ ] Tests: FMG converges to the same solution as N V-cycles in fewer total sweeps on a Poisson/Helmholtz benchmark; `smooth_abs`/`smooth_clamp` match the hard versions away from the kink and are finite-gradient at the kink (`jax.grad` is finite at `x = 0` / `x = x_min`).
- [ ] One-line mention in the solvers docs.

## Notes / non-goals

- **No required gap here.** Variable-coefficient Helmholtz on masked domains (the thing MASSH's multigrid is for) is **already implemented** in `finitevolx/_src/solvers/multigrid.py`. This issue is purely additive convenience; close as `wontfix` if it's not worth the surface area.

## References

- MASSH: https://github.com/leguillf/MASSH/tree/VarDyn
  - `models/model_qgsw/helmholtz_multigrid.py` — `MG_Helmholtz` (`V_cycle`, `FMG_Helmholtz`, `jacobi_smoothing`, `restrict`, `prolong`)
  - `models/model_qgsw/sw.py`, `reconstruction.py` — `smooth_abs`, `smooth_clamp`
- Existing finitevolx solver (the "already covered" baseline): `finitevolx/_src/solvers/multigrid.py`
- Background: Briggs, Henson & McCormick (2000), *A Multigrid Tutorial*, 2nd ed. (FMG, §3–4).
